chain/chunks/src/client.rs

```diff
@@ -104,6 +106,40 @@ impl ShardedTransactionPool {
         }
         reintroduced_count
     }
+
+    /// Migrate all of the transactions in the pool from the old shard layout to
+    /// the new shard layout.
+    /// It works by emptying the pools for old shard uids and re-inserting the
+    /// transactions back to the pool with the new shard uids.
+    pub fn reshard(&mut self, old_shard_layout: &ShardLayout, new_shard_layout: &ShardLayout) {
+        tracing::debug!(
+            target: "client",
+            old_shard_layout_version = old_shard_layout.version(),
+            new_shard_layout_version = new_shard_layout.version(),
+            "resharding the transaction pool"
+        );
+        debug_assert!(old_shard_layout != new_shard_layout);
+        debug_assert!(old_shard_layout.version() + 1 == new_shard_layout.version());
+
+        let mut transactions = vec![];
+
+        let old_shard_uids = old_shard_layout.get_shard_uids();
+        for old_shard_uid in old_shard_uids {
+            if let Some(mut iter) = self.get_pool_iterator(old_shard_uid) {
+                while let Some(group) = iter.next() {
+                    while let Some(tx) = group.next() {
+                        transactions.push(tx);
+                    }
+                }
+            }
+        }
```

Comment:
don't we need to do `self.tx_pools.clear()` before inserting new transactions?

Reply:
Good question, I don't think it is needed, but I may add it anyway just in case. Thanks for bringing that up. 
- The transaction pools handle de-duplication. 
- The iterator is draining so all transactions will be moved out of the old pools. 
- As of this PR the transactions pools are indexed by ShardUId. If any transactions are added there while the resharding is happening, it's ok. 
- One worry is what happens if transactions are added to the pools for old shard layout ids after the resharding. Those could be stuck there forever. I don't think it's a massive issue so I won't be fixing it here but let me add a todo.